src/controllers/publish.ts

```diff
@@ -295,10 +279,223 @@ export const taskList = async (req: Request, res: Response, next: NextFunction)
+        errors = [{ field: 'preview', message: { key: 'errors.preview.failed_to_get_preview' } }];
+    }
+    res.render('publish/cube-preview', { ...previewData, dataset, pagination, errors });
+};
+
+export const downloadAsCSV = async (req: Request, res: Response, next: NextFunction) => {
+    const dataset = singleLangDataset(res.locals.dataset, req.language);
+    const revision = res.locals.revision;
+    const fileStream = await req.swapi.getRevisionCubeCSV(dataset.id, revision.id);
+    res.writeHead(200, {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-Type': 'text/csv; charset=utf-8',
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-disposition': `attachment;filename=${revision.id}.csv`
+    });
+    const readable: Readable = Readable.from(fileStream);
+    readable.pipe(res);
+};
+
+export const downloadAsParquet = async (req: Request, res: Response, next: NextFunction) => {
+    const dataset = singleLangDataset(res.locals.dataset, req.language);
+    const revision = res.locals.revision;
+    const fileStream = await req.swapi.getRevisionCubeParquet(dataset.id, revision.id);
+    res.writeHead(200, {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-Type': 'application/vnd.apache.parquet',
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-disposition': `attachment;filename=${revision.id}.parquet`
+    });
+    const readable: Readable = Readable.from(fileStream);
+    readable.pipe(res);
+};
+
+export const downloadAsExcel = async (req: Request, res: Response, next: NextFunction) => {
+    const dataset = singleLangDataset(res.locals.dataset, req.language);
+    const revision = res.locals.revision;
+    const fileStream = await req.swapi.getRevisionCubeExcel(dataset.id, revision.id);
+    res.writeHead(200, {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-Type': 'application/vnd.ms-excel',
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-disposition': `attachment;filename=${revision.id}.xlsx`
+    });
+    const readable: Readable = Readable.from(fileStream);
+    readable.pipe(res);
+};
+
+export const downloadAsDuckDb = async (req: Request, res: Response, next: NextFunction) => {
+    const dataset = singleLangDataset(res.locals.dataset, req.language);
+    const revision = res.locals.revision;
+    const fileStream = await req.swapi.getRevisionCube(dataset.id, revision.id);
+    res.writeHead(200, {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-Type': 'application/octet-stream',
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-disposition': `attachment;filename=${revision.id}.duckdb`
+    });
+    const readable: Readable = Readable.from(fileStream);
+    readable.pipe(res);
+};
```

Comment:
As with the backend it feels like we could slim this down to one request handler and one api method e.g. something like

```ts
export const download = async (req: Request, res: Response, next: NextFunction) => {
    const dataset = singleLangDataset(res.locals.dataset, req.language);
    const revision = res.locals.revision;
    const format = req.query.format;

    const fileStream = await req.swapi.getRevisionCube(dataset.id, revision.id, format);
    res.writeHead(200, getContentHeaders(format));
    const readable: Readable = Readable.from(fileStream);
    readable.pipe(res);
};
```